solver/llbsolver/ops/exec.go

```diff
@@ -331,30 +331,26 @@ func (g *cacheRefGetter) getRefCacheDirNoCache(ctx context.Context, key string,
 	return mRef, nil
 }
 
-func (e *execOp) getSSHMountable(ctx context.Context, m *pb.Mount) (cache.Mountable, error) {
-	sessionID := session.FromContext(ctx)
-	if sessionID == "" {
-		return nil, errors.New("could not access local files without session")
-	}
-
-	timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
-	defer cancel()
-
-	caller, err := e.sm.Get(timeoutCtx, sessionID)
-	if err != nil {
-		return nil, err
-	}
-
-	if err := sshforward.CheckSSHID(ctx, caller, m.SSHOpt.ID); err != nil {
-		if m.SSHOpt.Optional {
-			return nil, nil
-		}
-		if grpcerrors.Code(err) == codes.Unimplemented {
-			return nil, errors.Errorf("no SSH key %q forwarded from the client", m.SSHOpt.ID)
+func (e *execOp) getSSHMountable(ctx context.Context, m *pb.Mount, g session.Group) (cache.Mountable, error) {
+	var caller session.Caller
+	err := e.sm.Any(ctx, g, func(ctx context.Context, _ string, c session.Caller) error {
+		if err := sshforward.CheckSSHID(ctx, c, m.SSHOpt.ID); err != nil {
+			if m.SSHOpt.Optional {
+				return nil
+			}
+			if grpcerrors.Code(err) == codes.Unimplemented {
+				return errors.Errorf("no SSH key %q forwarded from the client", m.SSHOpt.ID)
+			}
+			return err
 		}
+		caller = c
+		return nil
+	})
+	if err != nil {
 		return nil, err
 	}
-
+	// because ssh socket remains active, to actually handle session disconnecting ssh error
+	// should restart the whole exec with new session
```

Comment:
Can you expand on this? I don't understand.

Reply:
If we have 2 builds both running the same exec() with ssh mounted the ssh remains active for the whole duration of the exec. So if one session goes away there is no way to switch this ssh socket to another session as it might be in unknown state. Atm we ignore this and only validate session works when exec starts. But if we would want to handle this case then when connection would drop from ssh, we could just restart the whole exec with the new session. If you look at the "local source" implementation now, then this is what I do there. If a transfer fails it will check if we can attempt a new transfer from another session before failing the build.